Wip/mail

Sturmwoelfin opened the pull request:
### What does this PR do ?

This PR implements the email verification for a user. For this, it uses the field email_nonce in the database to store a uuid until the email is validated. 
The changes occur in  the modules ```MailModule``` and ```AuthModule```.

#### Functionalities

- [x] send email to user when creating account
- [x]  user can follow link to verify email address
- [x] user can request password reset
- [x] user can reset password  

A user is considered verified / active when the email_nonce of his account === null.

#### Test

To test mailhog locally, you can simply specify 127.0.0.1 as host & mailer_from.

Deploy local mail server : 
```docker compose up```

Start api : 
```npm run start```

With Postman (or other) send a request to http://<HOST>:<API_PORT>/auth/register
Then you need to check on http://localhost:8025 where Mailhog is located. Refresh the page and you should see a new email. Follow the link indicated and you should be redirected to another page where your email will automatically be verified.

#### TODO

When we go into production, we can't continue using local emails. This setup works fine with gmail or other mail servers but we would need to select one common to everyone.

Commit:
feat: added password reset template

sea-gull-diana commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,66 @@
+  async login(credentials: LoginUserDto): Promise<AccessToken> {
+    const user = await this.usersService.findOneUnsanitized({ email: credentials.email });
+    if (user.email_nonce !== null) {
+      throw new UnauthorizedException();
```
Maybe use HttpException with a clear error instead of just Unauthorized here.
```suggestion
      throw new HttpException('Email address not confirmed', HttpStatus.UNAUTHORIZED);
```

sea-gull-diana commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,66 @@
+  // Sign user in by email and password
+  async login(credentials: LoginUserDto): Promise<AccessToken> {
+    const user = await this.usersService.findOneUnsanitized({ email: credentials.email });
+    if (user.email_nonce !== null) {
```
There's a piece of code missing where an existence of `user` is checked. Add it to your line please:
```suggestion
    if (!user || user.email_nonce !== null) {
```

sea-gull-diana commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,66 @@
+    }
+    const passwordHash = user.password;
+    const isCorrectPassword = await bcrypt.compare(credentials.password, passwordHash);
+    if (!isCorrectPassword || !passwordHash) {
```
Probably some stuff to delete coming from my commit, my bad. It is impossible to have a user without a password, no need to check for it.
```suggestion
    if (!isCorrectPassword) {
```

sea-gull-diana commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,66 @@
+  // verify email token
+  async confirmEmail(token: string): Promise<Boolean> {
+    let sanitizedUser = await this.usersService.findOne({ email_nonce: token });
+    if (sanitizedUser == null) {
```
```suggestion
    if (!sanitizedUser) {
```

sea-gull-diana commented on src/auth/guards/auth.guard.ts:
```diff
@@ -0,0 +1,48 @@
+    UnauthorizedException,
+  } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+  import { JwtService } from '@nestjs/jwt';
```
```suggestion
import { JwtService } from '@nestjs/jwt';
```

sea-gull-diana commented on src/auth/guards/auth.guard.ts:
```diff
@@ -0,0 +1,48 @@
+  } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+  import { JwtService } from '@nestjs/jwt';
+  import { Request } from 'express';
```
```suggestion
import { Request } from 'express';
```

sea-gull-diana commented on src/auth/guards/auth.guard.ts:
```diff
@@ -0,0 +1,48 @@
+  import { Request } from 'express';
+import { IS_PUBLIC_KEY } from '../decorators/public.decorator';
+  
+  @Injectable()
```
```suggestion
@Injectable()
```

sea-gull-diana commented on src/mail/mail.service.ts:
```diff
@@ -27,14 +37,38 @@ export class MailService {
     }
 
-    async sendUserConfirmation(user: User, token: string) {
+    async sendUserConfirmation(email: string, token: string) {
```
It would be good to define what this function returns.
```suggestion
    async sendUserConfirmation(email: string, token: string): Promise<boolean> {
```

sea-gull-diana commented on src/mail/mail.service.ts:
```diff
@@ -27,14 +37,38 @@ export class MailService {
-        await this.mailerService.sendMail({
-            to: user.email,
+
+        return await this.mailerService.sendMail({
```
Normally, you shouldn't do both `await` and `then/catch` right? You use Promises with `await` to avoid using `then`. 
Maybe just do `this.mailerService.sendMail(... then(... return..., catch(... return...`, without `return await`?

Sturmwoelfin replied:
Good point, I'll change that !

sea-gull-diana commented on src/users/users.service.ts:
```diff
@@ -1,36 +1,59 @@
+  }
+
+  async validateEmailNonce(uuid: string): Promise<boolean> {
+    await this.prisma.user.update({
```
I suggest using a try-catch block here, since we don't won't to return true if user update fails for some reason.

Sturmwoelfin replied:
try catch or more something like 
```js
let user = await this.prisma.user.update(); 
if (!user) {
return false
}
return true;
```

sea-gull-diana replied:
I'm not sure if prisma will just return null on error. If it throws an uncaught exception, it may crush our API. I found this code example in documentation: https://www.prisma.io/docs/concepts/components/prisma-client/handling-exceptions-and-errors

sea-gull-diana commented on src/utils/response.dto.ts:
```diff
@@ -0,0 +1,35 @@
+import { IResponse } from "./response.interface";
```
This file needs to be deleted. We do not use Response classes.

sea-gull-diana commented on src/utils/response.interface.ts:
```diff
@@ -0,0 +1,7 @@
+export interface IResponse {
```
This file needs to be deleted too XD

sea-gull-diana reviewed (changes requested):
Hi! Thanks for your PR :) It looks good to me :+1:  There's a problem with Git reapplying all of the Authentication PR changes into yours, but I'm not sure we can do anything with it.

I found a bunch of my own mishaps, that you could correct in your PR (see comments) ;)

Also there is some not deleted stuff because of merge conflicts I mentioned in my suggestions as well as a couple of ideas to improve the code.

Commit:
feat: mailservice for password request

Commit:
feat: added email validation

Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

Commit:
feat: added email validation

Commit:
feat: added email validation

Commit:
feat: added email validation

Sturmwoelfin committed:
fix: better Exception when email not validated fot login

Co-authored-by: Dziyana <67862638+sea-gull-diana@users.noreply.github.com>

Sturmwoelfin committed:
fix: better Exception when email not validated fot login

Co-authored-by: Dziyana <67862638+sea-gull-diana@users.noreply.github.com>

Sturmwoelfin committed:
fix: user verification

Co-authored-by: Dziyana <67862638+sea-gull-diana@users.noreply.github.com>

Sturmwoelfin committed:
fix: user verification

Co-authored-by: Dziyana <67862638+sea-gull-diana@users.noreply.github.com>

Sturmwoelfin committed:
fix: code refactor

Co-authored-by: Dziyana <67862638+sea-gull-diana@users.noreply.github.com>
Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

Sturmwoelfin committed:
fix: code amelioration

Co-authored-by: Dziyana <67862638+sea-gull-diana@users.noreply.github.com>
Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

Sturmwoelfin committed:
fix: function return type defined

Co-authored-by: Dziyana <67862638+sea-gull-diana@users.noreply.github.com>

Commit:
fix: code refactoring

Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

Commit:
fix: code refactoring

Commit:
feat: added unique constraint

Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

Commit:
feat: added unique constraint

Commit:
fix: email uuid is automatically created in the db

Commit:
fix: email uuid is automatically created in the db

Commit:
fix: email uuid is automatically created in the db

Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

Commit:
Database refactor: uuids and uniques (#11)

Refactor the database by changing all integer IDs into UUIDs, except for
the SSH keys, who benefit from staying in "normal" IDs; Remove the
unique constraints where they shouldn't have been, in order to avoid
problems; Add password nonce to enable generation of a "forgotten
password" email.

BREAKING CHANGE

Commit:
Database refactor: uuids and uniques (#11)

Refactor the database by changing all integer IDs into UUIDs, except for
the SSH keys, who benefit from staying in "normal" IDs; Remove the
unique constraints where they shouldn't have been, in order to avoid
problems; Add password nonce to enable generation of a "forgotten
password" email.

BREAKING CHANGE

Commit:
Database refactor: uuids and uniques (#11)

Refactor the database by changing all integer IDs into UUIDs, except for
the SSH keys, who benefit from staying in "normal" IDs; Remove the
unique constraints where they shouldn't have been, in order to avoid
problems; Add password nonce to enable generation of a "forgotten
password" email.

BREAKING CHANGE

Commit:
Database refactor: uuids and uniques (#11)

Refactor the database by changing all integer IDs into UUIDs, except for
the SSH keys, who benefit from staying in "normal" IDs; Remove the
unique constraints where they shouldn't have been, in order to avoid
problems; Add password nonce to enable generation of a "forgotten
password" email.

BREAKING CHANGE

mfernd commented on src/users/users.service.ts:
```diff
@@ -9,38 +9,56 @@ import * as bcrypt from 'bcrypt';
+    try {
+      await this.prisma.user.update({
+        where: {
+          emailNonce: uuid,
```
`emailNonce` is not unique in the schema, we can not use it in the `where` fields

sea-gull-diana committed:
feat: use passport-jwt for authentication (#9)

Modifies user authentication by using passport-jwt strategy.

Add refresh tokens and possibly API tokens (for CLI login).

---------

Co-authored-by: BioTheWolff <biothewolff@gmx.fr>

Commit:
fix: removed double unique index

Commit:
fix(auth): fix login method and user validation

mfernd reviewed (dismissed):
runs fine for me :ok_person:

Commit:
style: revert changes that could have been left

Commit:
style: fix style for password reset email template

Commit:
refactor(auth|mail): add needed changes

sea-gull-diana committed:
Merge branch 'main' into wip/mail

Commit:
feat: password reset implementation

Commit:
fix: commentary

noetarbouriech reviewed (dismissed):
lgtm

BioTheWolff commented on src/auth/auth.controller.ts:
```diff
@@ -31,4 +31,41 @@ export class AuthController {
+    // POST /auth/password/reset
+    // This path lets the user reset his password
+    @Public()
+    @Post('password/reset')
+    async resetPassword(@Query('token') token: string, @Body() passwordResetDto: PasswordResetDto): Promise<string> {
```
Using `@Param()` is preferable to `@Query()` when possible. It uses the request parameters instead of looking at the query.

```suggestion
    @Post('password/reset/:token')
    async resetPassword(@Param('token') token: string, @Body() passwordResetDto: PasswordResetDto): Promise<string> {
```

BioTheWolff commented on src/auth/auth.controller.ts:
```diff
@@ -31,4 +31,41 @@ export class AuthController {
+    // POST /auth/confirm/:token
+    // This path lets the user confirm his email
+    @Public()
+    @Post('confirm')
+    async confirmEmail(@Query('token') token: string): Promise<string> {
```
Using `@Param()` is preferable to `@Query()` when possible. Should be useful to enforce calling a route with a given parameter

```suggestion
    @Post('confirm/:token')
    async confirmEmail(@Param('token') token: string): Promise<string> {
```

BioTheWolff commented on src/auth/dto/password-reset.dto.ts:
```diff
@@ -0,0 +1,13 @@
+import{ IsEmail, IsNotEmpty } from 'class-validator';
+
+export class PasswordResetDto {
+    @IsEmail()
+    @IsNotEmpty()
+    email: string;
+
```
We don't need the email in the password reset. We are assuming the user is redirected from their email address.

```suggestion
```

BioTheWolff commented on src/mail/templates/password-reset.hbs:
```diff
@@ -0,0 +1,10 @@
+<h2>Password reset</h2>
+<p>Please click the link below to reset your password:</p>
+<p>
+    <a href="{{ url }}">Confirm</a>
```
Make the template match with the string template in `mail.service.ts`

```suggestion
    <a href="{{ url }}">Reset password</a>
```

BioTheWolff commented on src/auth/auth.service.ts:
```diff
@@ -55,4 +66,43 @@ export class AuthService {
+    try {
+      let user = await this.usersService.findOneUnsanitized({email: passwordRequestDto.email})
+      if (user) {
+        let passwordNonce = uuidv4()
+        await this.usersService.updatePasswordNonce(passwordNonce, user.id);
```
Generating the password nonce should remain responsibility of the Users service. Please see comment regarding this on the `users.service.ts` file.

BioTheWolff commented on src/users/users.service.ts:
```diff
@@ -10,38 +10,91 @@ import { v4 as uuidv4 } from 'uuid';
+
   }
 
+  async updatePasswordNonce(token: string, id: string): Promise<boolean> {
+    try {
+      await this.prisma.user.update({
+        where: {
+          id: id,
+        },
+        data: {
+          passwordNonce: token
+        }
+      })
+      return true;
+    } catch (err) {
+      return false;
+    }
+  }
```
I recommend not accepting strings from outside for the regenerated token.

```suggestion
  async regeneratePasswordNonce(id: string): Promise<boolean> {
    try {
      await this.prisma.user.update({
        where: {
          id: id,
        },
        data: {
          passwordNonce: uuidv4()
        }
      })
      return true;
    } catch (err) {
      return false;
    }
  }
```

BioTheWolff commented on src/auth/auth.service.ts:
```diff
@@ -6,24 +6,35 @@ import * as bcrypt from 'bcrypt';
-    return await this.usersService.create(createUserDto);
+    let user = await this.usersService.create(createUserDto);
+    
+    if (!await this.mailService.sendUserConfirmation(createUserDto.email, user.emailNonce)) {
```
Make use of precedently generated `user`! It comes from the database, which is the source of truth. DTOs should only be used as long as the data is in transit towards the database.

```suggestion
    if (!await this.mailService.sendUserConfirmation(user.email, user.emailNonce)) {
```

BioTheWolff commented on src/auth/auth.service.ts:
```diff
@@ -55,4 +66,43 @@ export class AuthService {
+
+  // reset password
+  async passwordReset(token: string, passwordResetDto: PasswordResetDto) : Promise<boolean> {
+    if (passwordResetDto.password !== passwordResetDto.passwordVerification) {
+      return false;
+    }
```
Optional: could be done within the DTO using a custom decorator. This is not required for the PR to pass

BioTheWolff reviewed (changes requested):
Looking good overall, just a few things to change:
- The `@Query()` should always be replaced by `@Param()` if they are required within the route, or if they are always included; Query parameters should remain used as little as possible, as the usecase is very niche;
- The Users service should keep responsibility of regenerating the password nonce token
- Some other minor fixes have been suggested

Commit:
fix: refactoring

Commit:
feat: add match other validator and its use in password reset

Commit:
fix: added Password max length

Commit:
fix: added pw verification for create-user dto

Commit:
fix: commentary

Commit:
fix: removed unnecessary password check

BioTheWolff commented on src/auth/dto/password-reset.dto.ts:
```diff
@@ -0,0 +1,13 @@
+
+    @IsNotEmpty()
+    @MatchesWithProperty(PasswordResetDto, o => o.password)
+    passwordVerification: string;
```
This should implement the password regex, like the password that is set during account creation. See `create-user.dto.ts` file for reference

```ts
@Matches(PASSWORD_REGEX, {
        message: 'Password should be minimum eight character long and have at least one uppercase letter, one lowercase letter, one number and one special character.'
    })
```

BioTheWolff reviewed (changes requested):
Just one more change to do! This is for added security, and to ensure password reset doesn't break the initially enforced password rules

Commit:
fix: password verification on reset

Commit:
fix: added constraints to password verification

BioTheWolff reviewed (approved):
Looks good to me!!

mfernd reviewed (approved):
runs fine for me :ok:
